### DISCUSSION TITLE
Would there be interest in including RPM packaging as part of the code/GH Actions?

### DISCUSSION THREAD

**Discussion by alexpdp7** [seaweedfs/seaweedfs#5248] [Ideas]
Hi,

I'm mulling deploying SeaweedFS for personal purposes. I prefer to deploy software with dnf on EL9. [I maintain a few RPMs for this](https://copr.fedorainfracloud.org/coprs/koalillo/).

I mostly work "outside" the target project, keeping build materials in a separate Git repo. This works well, but it's a bit more of effort, because I need to track upstream distributions and release RPMs whenever there's a new release of the target software.

There is the possibility of rolling the packaging inside the main project. As SeaweedFS is using GH Actions, I think I could extend the release workflows to build RPMs and push them to a public repo.

Would there be interest in accepting PRs for the latter option? As long as I'd use SeaweedFS I think I could take care of maintenance, but I cannot offer "professional" support. Although in the case of such a self-contained software, I think it wouldn't be a ton of work anyway...

Thoughts?

Álex

**Comment by chrislusf** [comment]
Totally yes!

**Comment by alexpdp7** [comment]
You mean, it would be possible to merge it in the main repo? If so, I'll see if I work on that, thanks.

**Comment by alexpdp7** [comment]
I have drafted <https://github.com/alexpdp7/seaweedfs-rpm>.

This is the most minimum RPM that works. For the moment, I kept it out of the main repo (or out of a fork, rather), because I didn't want to work on building the command. The `rpm.py` script fetches a release from GitHub to obtain a binary, and builds an RPM with just that.

This is a reduced version of <https://github.com/miniflux/v2/tree/main/packaging/rpm>, which I think is a nice example of building an RPM from an existing binary. The Miniflux RPM also includes creating a user and a systemd unit- I'm not creating the systemd unit yet because I have some doubts about how to proceed.

This package would *NOT* be viable for Fedora or RHEL repos. The "right" way to do an RPM is to use the [go2rpm](https://pagure.io/GoSIG/go2rpm/), which requires packaging the dependencies. If including SeaweedFS in RHEL/Fedora is the goal, then a different approach is needed.

It's a bit bizarre to use Python and podman to build (a shell script and Docker would be enough), but hopefully the code is clear enough that is useful if someone wants to try a different approach.

**Comment by alexpdp7** [comment]
There's also the subject of publishing the RPM to a yum repo. I see two main options:

* COPR: is an official Fedora resource. It is nice, because it can do builds- even multiarch builds. I use this for packages I only want in RHEL: [1](https://github.com/alexpdp7/vaultwarden-rpm), [2](https://src.fedoraproject.org/fork/koalillo/rpms/nextcloud).
* Cloudsmith: offers free repos, and supports .deb on top of .rpm. [I use it for a package I build that has Debian and RHEL versions](https://github.com/alexpdp7/ragent/blob/master/.github/workflows/ci.yml#L105).

**Reply by alexpdp7** [reply]
OK, I have added a bare-minimum systemd unit, created a package, and uploaded it to a COPR.

The instructions at:

https://github.com/alexpdp7/seaweedfs-rpm/tree/main?tab=readme-ov-file#install

Should work in RHEL 9 and derivatives (I just tested it in CentOS 9 Stream).

**Reply by alexpdp7** [reply]
@chrislusf what would you consider as the "bare minimum" to try to mainline the RPM into your setup? (edit: if you wish to do that, of course!!)

I just pushed an update for the RPM for 3.64 and it worked well. Right now the package only provides the most minimal setup- but it might be the right thing? (install the package, enable the service, you are up and running. You can always add more customizations...)

I kinda like that if you deploy the RPM and set up `dnf-automatic`, you are automatically patched without effort.
